components/org.wso2.carbon.identity.application.authenticator.google/src/main/java/org/wso2/carbon/identity/application/authenticator/google/GoogleOAuth2Authenticator.java

```diff
@@ -51,13 +69,191 @@ public class GoogleOAuth2Authenticator extends OpenIDConnectAuthenticator {
         }
     }
 
+    @Override
+    public boolean canHandle(HttpServletRequest request) {
+        /*
+          Validity of the
+          1. CSRF cookies
+          2. JWT token
+          decide the ability of handling the authentication request in Google One Tap flow.
+         */
+        if (isOneTapEnabled(request)) {
+            boolean validCookies = validateCSRFCookies(request);
+            if (validCookies) {
+                return validateJWTFromGOT(request);
+            }
+        }
+        return super.canHandle(request);
+    }
+
+    /**
+     * This function validates the JWT token by its content using Google libraries.
+     *
+     * @param request Authentication request with JWT token.
+     * @return Validity of the returned JWT token returned via Google One Tap.
+     */
+    private boolean validateJWTFromGOT(HttpServletRequest request) {
+
+        ExternalIdPConfig externalIdPConfig = null;
+        try {
+            String authenticatorName = request.getParameter(IDP);
+            String tenantDomain = String.valueOf(request.getAttribute(TENANT_DOMAIN));// Returns "null" for null
+
+            externalIdPConfig = ConfigurationFacade.getInstance()
+                    .getIdPConfigByName(authenticatorName, tenantDomain);
+        } catch (Exception e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Error in identifying the authenticator.", e);
+            }
+        }
+        if (externalIdPConfig == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Cannot identify the authenticator.");
+            }
+            return false;
+        }
+        String clientId = FrameworkUtils.getAuthenticatorPropertyMapFromIdP(externalIdPConfig, getName())
+                .get(OIDCAuthenticatorConstants.CLIENT_ID);
+
+        String idTokenString = request.getParameter(CREDENTIAL);
+        //Verifying the ID token.
+        ApacheHttpTransport transport = new ApacheHttpTransport();
+        GsonFactory jsonFactory = new GsonFactory();
+
+        /*
+          Specify the CLIENT_ID of the app that accesses the backend:
+          Or, if multiple clients access the backend:
+          .setAudience(Arrays.asList(CLIENT_ID_1, CLIENT_ID_2, CLIENT_ID_3))
+         */
+        GoogleIdTokenVerifier verifier = new GoogleIdTokenVerifier.Builder(transport, jsonFactory)
+                .setAudience(Collections.singletonList(clientId))
+                .build();
+        GoogleIdToken idToken = null;
+        try {
+            idToken = verifier.verify(idTokenString);
+        } catch (GeneralSecurityException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("In-secured JWT returned from Google One Tap.", e);
+            }
+        } catch (IOException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Exception while validating the JWT returned from Google One Tap.", e);
```

Comment:
Why this is a debug log?

Reply:
Fixed f806cee9c82a8896103acc5729765b340d3190c1